pkg/hubble/metrics/dynamic_flow_processor.go

```diff
@@ -0,0 +1,151 @@
+
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DynamicFlowProcessor represents instance of hubble exporter with dynamic
+// configuration reload.
+type DynamicFlowProcessor struct {
+	logger  logrus.FieldLogger
+	watcher *metricConfigWatcher
+	// Protects against deregistering metric handlers while ProcessFlow is executing, or concurrent config reloads.
+	mutex    lock.RWMutex
+	Metrics  []api.NamedHandler
+	registry *prometheus.Registry
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *DynamicFlowProcessor) OnDecodedFlow(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, d.Stop()
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var errs error
+	if d.Metrics != nil {
+		errs = api.ExecuteAllProcessFlow(ctx, flow, &d.Metrics)
+	}
+
+	if errs != nil {
+		d.logger.WithError(errs).Error("Failed to ProcessFlow in metrics handler")
+	}
+	return false, errs
+}
+
+// Stop stops configuration watcher  and all deinitializes all metric handlers.
+func (d *DynamicFlowProcessor) Stop() error {
+	d.watcher.Stop()
+
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	var errs error
+	for _, h := range d.Metrics {
+		errs = errors.Join(errs, h.Handler.Deinit(d.registry))
+	}
+
+	return errs
+}
+
+// NewDynamicFlowProcessor creates instance of dynamic hubble flow exporter.
+func NewDynamicFlowProcessor(reg *prometheus.Registry, logger logrus.FieldLogger, configFilePath string) *DynamicFlowProcessor {
+	dynamicFlowProcessor := &DynamicFlowProcessor{
+		logger:   logger,
+		registry: reg,
+	}
+	watcher := NewMetricConfigWatcher(configFilePath, dynamicFlowProcessor.onConfigReload)
+	dynamicFlowProcessor.watcher = watcher
+	return dynamicFlowProcessor
+}
+
+func (d *DynamicFlowProcessor) onConfigReload(ctx context.Context, hash uint64, config api.Config) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	var newHandlers []api.NamedHandler
+	metricNames := config.GetMetricNames()
+
+	curHandlerMap := make(map[string]*api.NamedHandler)
+	if d.Metrics != nil {
+		for _, m := range d.Metrics {
+			curHandlerMap[m.Name] = &m
+		}
+
+		// Unregister handlers not present in the new config.
+		// This needs to happen first to properly check for conflicting plugins later during registration.
+		for _, m := range d.Metrics {
+			if _, ok := metricNames[m.Name]; !ok {
+				h := curHandlerMap[m.Name]
+				err := h.Handler.Deinit(d.registry)
+				if err != nil {
+					d.logger.WithField("name", m.Name).WithError(err).Error("Deinit failed for handler")
+				}
+				delete(curHandlerMap, m.Name)
+			}
+		}
+	}
+
+	for _, cm := range config.Metrics {
+		// Existing handler matches new config entry:
+		//   no-op, if config unchanged;
+		//   update handler config, if changed.
+		if m, ok := curHandlerMap[cm.Name]; ok {
+			if reflect.DeepEqual(*m.MetricConfig, *cm) {
+				continue
+			} else {
```

Comment:
This else clause is superfluous, removing it would make the code more readable. In general, we should try to keep the code path at minimal indentation.

Reply:
Removed 2 levels of indentation - the `if` inside this `else`was duplicating the outer `if.`